crates/dekoder/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+            ))? as f64
+        };
+        let nf = yml["nf"]
+            .as_i64()
+            .ok_or(EKOError::KeyError("because failed to read nf".to_owned()))?;
+        Ok(Self { scale, nf })
+    }
+}
+
+impl PartialEq for EvolutionPoint {
+    /// (Protected) comparator.
+    fn eq(&self, other: &Self) -> bool {
+        self.nf == other.nf && approx_eq!(f64, self.scale, other.scale, ulps = EP_CMP_ULPS)
+    }
+}
+
+impl Eq for EvolutionPoint {}
+
+/// EKO output
+pub struct EKO {
+    /// Working directory
+    path: PathBuf,
+    /// final operators
+    operators: inventory::Inventory<EvolutionPoint>,
+}
+
+/// Operators directory.
+const DIR_OPERATORS: &str = "operators/";
+
+impl EKO {
+    /// Check our working directory is safe.
+    fn check(&self) -> Result<()> {
+        let path_exists = self.path.try_exists().is_ok_and(|x| x);
+        if !path_exists {
+            return Err(EKOError::NoWorkingDir);
+        }
+        Ok(())
+    }
+
+    /// Remove the working directory.
+    pub fn destroy(&self) -> Result<()> {
+        self.check()?;
+        Ok(remove_dir_all(&self.path)?)
+    }
+
+    /// Write content back to an archive and destroy working directory.
+    pub fn write_and_destroy(&self, dst: PathBuf) -> Result<()> {
+        self.write(dst)?;
+        self.destroy()?;
+        Ok(())
+    }
+
+    /// Write content back to an archive.
+    pub fn write(&self, dst: PathBuf) -> Result<()> {
+        self.check()?;
+        // create writer
+        let dst_file = File::create(&dst)?;
+        let dst_file = BufWriter::with_capacity(128 * 1024, dst_file);
+        let mut ar = tar::Builder::new(dst_file);
+        // do it!
+        ar.append_dir_all(".", &self.path)?;
+        Ok(())
+    }
+
+    /// Extract tar file from `src` to `dst`.
+    pub fn extract(src: PathBuf, dst: PathBuf) -> Result<Self> {
+        let mut ar = tar::Archive::new(File::open(&src)?);
+        ar.unpack(&dst)?;
+        Self::load_opened(dst)
+    }
+
+    /// Load an EKO from a directory `path` (instead of tar).
+    pub fn load_opened(path: PathBuf) -> Result<Self> {
+        let mut operators = inventory::Inventory {
+            path: path.join(DIR_OPERATORS),
+            keys: HashMap::new(),
+        };
+        operators.load_keys()?;
+        let obj = Self { path, operators };
+        obj.check()?;
+        Ok(obj)
+    }
+
+    /// List available evolution points.
+    pub fn available_operators(&self) -> Vec<&EvolutionPoint> {
+        self.operators.keys()
+    }
+
+    /// Check if the operator at the evolution point `ep` is available.
+    pub fn has_operator(&self, ep: &EvolutionPoint) -> bool {
+        self.operators.has(ep)
+    }
+
+    /// Load the operator at the evolution point `ep` from disk.
+    pub fn load_operator(&self, ep: &EvolutionPoint, op: &mut Operator) -> Result<()> {
+        self.check()?;
+        self.operators.load(ep, op)?;
+        Ok(())
+    }
```

Comment:
Which is the advantage of loading into the `Operator`?

If you are not trying to save memory allocations (the expensive operation here is loading, more than allocating memory), I'd suggest you to give up on return-by-argument (that's really Fortran style...) and just return the operator as a return value.

Reply:
as said [on top](https://github.com/NNPDF/eko/pull/400#issue-2461037962) I'd like to keep it this way, because then we don't need to deal with the memory problem here, but it is the users responsibility. And the crucial thing is that I don't hold the object and so I have to take it by reference. Yes, this is Fortran style and that's the price we have to pay to get the memory management "for free". I don't know yet how this will play out in the Python interface - maybe we have to copy it to the other side and delegate again it's management to the user

Reply:
You don't need a separate object to do that. Even without references, Rust has move semantics by default (the C++ move), so if you return the object, the control is flowing to the caller. And if you pass the object by value to another function, the control is passed inside.
The ownership rule is made exactly for this: move memory without copying, avoiding multiple references to cause memory issues.

You can just return, and stay safe and efficient.

> I don't know yet how this will play out in the Python interface - maybe we have to copy it to the other side and delegate again it's management to the user

This is more complicated. For as long as the manipulation is in Rust, you can return an object with the hidden reference to Python, and just call the Rust API to manipulate memory. This is exactly what NumPy is doing in C (it's never giving back a Python list, unless forced with `.tolist()` - but it's constructing a rich API to manipulate its own objects).
However, if you want a NumPy array you will to "translate" it somehow, from a Rust ndarray one. This could be done efficiently, as Python has standardized the concept of memory buffers (essentially, you describe the structure of your buffer, and pass a reference to it to another object). But you don't want to do this manually - just hope that the Rust `numpy` crate is doing it for you.